feat: quaternion synchronization, delta position, and NetworkTransform bandwidth optimizations

NoelStephensUnity opened the pull request:
This update provides users with optional bandwidth optimizations (at the cost of precision) of up to 50% reduction in bandwidth costs per `NetworkTransform` instances. It also includes several changes that address issues related to rotation and parented `NetworkTransform`s through the enabling of quaternion synchronization.

## `NetworkTransform`  Additions and Changes
- `Quaternion` Synchronization:  Added the ability to synchronize by quaternions (_full precision_). Using quaternion synchronization is recommended for scenarios where multiple axis of rotation can change and the transform in question is within a parent-child hierarchy. When enabled, per axis rotation flags are no longer visible in the editor as the entire quaternion is synchronized when any change in rotation occurs.
- `Quaternion` Compression:  Added the ability to compress quaternions (_requires `Quaternion` Synchronization to be enabled_). Reduces the cost of quaternion synchronization down to 4 bytes at the cost of additional loss in precision (a bit less precise than half float precision).
- Half Float Precision: Added the ability to enable half float precision which reduces payload size per instance by ~50% at the cost of a decrease in precision.
- Delta Position Synchronization: When using half float precision `NetworkTransform` uses delta position synchronization to assure any loss in precision is recovered/handled in the next state update.
- Added additional virtual methods to be used in prediction calculation scenarios where knowing the exact values being updated are required. This addition includes changing `NetworkTransformState` to public.
- Added a bandwidth optimization related property, `SynchronizeTransform`,  to `NetworkObject` that allows users to disable the automatic synchronization of the associated transform to reduce the total bandwidth cost of spawning and late joining client synchronization. This is particularly useful if the `NetworkObject` is used more for management related tasks and has no spatial synchronization needs.

[Internal Companion Project](https://github.cds.internal.unity3d.com/noel-stephens/RNSM-NetworkTransform): This project demonstrates the bandwidth cost/savings based on which properties are set, the average precision loss over time (editor console log required), and can be used to visual the latency between the authoritative and non-authoritative instances.

## Fixes
[MTT-5252](https://jira.unity3d.com/browse/MTT-5252)
[MTT-5082](https://jira.unity3d.com/browse/MTT-5082)

**Related GitHub Issues Resolved:**
#2353
#2299
#1977

## Changelog

### Added
- Added: `NetworkTransform.UseHalfFloatPrecision` property that, when enabled, will use half float values for position, rotation, and scale. This yields a 50% bandwidth savings a the cost of precision. 
- Added: `NetworkTransform.UseQuaternionSynchronization` property that, when enabled, will synchronize the entire quaternion. 
- Added: `NetworkTransform.UseQuaternionCompression` property that, when enabled, will use a smallest three implementation reducing a full quaternion synchronization update to the size of an unsigned integer. 
- Added: `NetworkTransform.SlerpPosition` property that, when enabled along with interpolation being enabled, will interpolate using `Vector3.Slerp`. 
- Added: `BufferedLinearInterpolatorVector3` that replaces the float version, is now used by `NetworkTransform`, and provides the ability to enable or disable `Slerp`. 
- Added: `HalfVector3` used for scale when half float precision is enabled.
- Added: `HalfVector4` used for rotation when half float precision and quaternion synchronization is enabled.
- Added: `HalfVector3DeltaPosition` used for position when half float precision is enabled. This handles loss in position precision by updating only the delta position as opposed to the full position.
- Added: `NetworkTransform.GetSpaceRelativePosition` and `NetworkTransform.GetSpaceRelativeRotation` helper methods to return the proper values depending upon whether local or world space.
- Added: `NetworkTransform.OnAuthorityPushTransformState` virtual method that is invoked just prior to sending the `NetworkTransformState` to non-authoritative instances. This provides users with the ability to obtain more precise delta values for prediction related calculations. 
- Added: `NetworkTransform.OnNetworkTransformStateUpdated` virtual method that is invoked just after the authoritative `NetworkTransformState` is applied. This provides users with the ability to obtain more precise delta values for prediction related calculations. 
- Added: `NetworkTransform.OnInitialize` virtual method that is invoked after the `NetworkTransform` has been initialized or re-initialized when ownership changes. This provides for a way to make adjustments when `NetworkTransform` is initialized (i.e. resetting client prediction etc.)
- Added: `NetworkObject.SynchronizeTransform` property (default is `true`) that provides users with another way to help with bandwidth optimizations where, when set  to false, the `NetworkObject`'s associated transform will not be included when spawning and/or synchronizing late joining players.

### Changed
- Changed: `NetworkTransform` authority handles delta checks on each new network tick and no longer consumes processing cycles checking for deltas for all frames in-between ticks. 
- Changed: The `NetworkTransformState` structure is now public and now has public methods that provide access to key properties of the `NetworkTransformState` structure. 
- Changed: `NetworkTransform` interpolation adjusts its interpolation "ticks ago" to be 2 ticks latent if it is owner authoritative and the instance is not the server or 1 tick latent if the instance is the server and/or is server authoritative.

### Fixed
- Fixed: issue with a child's rotation rolling over when interpolation is enabled on a `NetworkTransform`. Now using half precision or full quaternion synchronization will always update all axis. 
- Fixed: issue where `NetworkTransform` was not setting the teleport flag when the `NetworkTransform.InLocalSpace` value changed. This issue only impacted `NetworkTransform` when interpolation was enabled. 

## Testing and Documentation
- Includes updates to existing integration tests.
- Includes a long-duration delta position manual test to check for precision drift over time.
- Includes edits to existing public API documentation.
- Requires additional public documentation updates.

NoelStephensUnity committed:
update

Work in progress.

NoelStephensUnity committed:
update

WIP
Closer but to get reasonable precision accuracy we utilize 6 bytes with only a 50% reduction in size.
Deltas cannot exceed +/- 128.0f (so 30x per second rate of change would be +/- 3840 unity space units per second)
The fault tolerance is +/- 0.20f

NoelStephensUnity committed:
update

Doing a comparison of the original vs the new implementation (precision relative).
Looks like 10 bit precision stored in a UINT and USHORT (6 bytes) is about the best combination between byte savings vs precision.

NoelStephensUnity committed:
update

Converted the original delta position compressor over to use 10 bit precision.
We lose 2 bytes of savings (like the adjustable precision version) but:
Deltas cannot exceed +/- 128.0f (so 30x per second rate of change would be +/- 3840 unity space units per second)
The decompressed value(s) fault tolerance is +/- 0.10f.
(50% reduction in size is the best precision vs compression ratio we can get with this approach)

NoelStephensUnity committed:
update

Improved the original DeltaPositionCompressor so it now has 11 bit precision for the smallest axial values.
Added some additional remarks about usage along with a precision accuracy table.
This version can handle +/-15,330 UWS units per second rate of change.

NoelStephensUnity committed:
update

These changes allow for increased precision at smaller delta position values.
It reaches its floor at +/- 0.001f delta (i.e. the minimum NetworkTransform position threshold)

NoelStephensUnity committed:
update

Renamed DeltaPositionCompressor to Vector3DeltaCompressor and made all necessary adjustments.
Integrated Vector3DeltaCompressor into NetworkTransform for position delta compression.

NoelStephensUnity committed:
Merge branch 'develop' into refactor/quaternion-synchronization-network-transform

NoelStephensUnity committed:
update

This gets NetworkTransform passing all tests with delta position compression disabled.
This fixes a few minor test related issues.

NoelStephensUnity committed:
update

Adjustments for the nested network transform test.
Preparing for the precision correction addition with delta position compression.

NoelStephensUnity committed:
update

Added adjustments to handle delta position precision loss.

NoelStephensUnity committed:
test manual

Added a manual test for long-duration testing of delta position compression with precision loss compensation.

NoelStephensUnity committed:
update

some additional minor tweaks.

NoelStephensUnity committed:
update

Forgot to check in the editor build settings.

NoelStephensUnity committed:
update

Now the non-authoritative side (client for this manual test) shows the absolute position of the client so when you stop motion it should be either the exact value on the server side or will have a very small loss in precision (depending upon when you stop moving).

NoelStephensUnity committed:
update

Adding a define for the delta position manual test and some cleaning up.

NoelStephensUnity committed:
update

updating the project settings again

NoelStephensUnity committed:
update

Adding quaternion synchronization without compression or precision loss compensation.
Some axis specific NetworkTransform integration tests fail due to the quaternion synchronizing all axis.
These tests will be adjusted once everything else (for quaternion compression and precision loss compensation) is in place.

NoelStephensUnity committed:
fix

This fixes an issue with the NestedNetworkTransform test.  The children NetworkTransforms were still using world space synchronization which caused issues when the scale changed.  When that was corrected, it then exposed a bug in the delta position implementation where the predicted client position was not taking local space into consideration.
Also making sure k_QuaternionSync is not cleared after updating on the authoritative side.

NoelStephensUnity committed:
test fix

Removing the sphere collider from the NetworkTransformOwnershipTests because the test basically ignores collision with other players and collision is not a factor of this test.  Since the spheres are effectively "doing nothing", it didn't make sense to try and add them when they weren't being used for anything and more importantly they were causing the client side to have skewed results when OwnerAuthoritativeTests was run under the HostStartsAsOwner pass.

NoelStephensUnity committed:
test updates

Temporarily adding updates to various NetworkTransform tests to account for current changes.
(primarily to run stand alone testrunner builds on consoles to verify current changes still work on them)

NoelStephensUnity committed:
update

Added HalfVector3
After doing some tests, it looks like using HalfVector3 yields the exact same size in bytes and actually has better over-all precision.

NoelStephensUnity committed:
update

All NetworkTransform integration tests pass using HalfVector3.
Still WIP, but looking like this could be the best path.

NoelStephensUnity committed:
update

Final tweaks to NetworkTransform using HalfVector3 and adjustments to the DeltaPositionNetworkTransform manual test to be able to compare both.

NoelStephensUnity committed:
tests update

Several adjustments to tests that take the HalfVector3 into consideration as well as some debug hooks for testing purposes.

NoelStephensUnity committed:
update

Added the quaternion compressor and the HalfVector4 (quat) implementation.
Updated some tests and NetworkTransform.

NoelStephensUnity committed:
update

This update:
- Increases the precision of HalfVector4 when using quaternions
- Adds UseHalfFloatPrecision property that when enabled will synchronize in half float values as well as synchronize the change in status on the non-authoritative sides.
- Includes some debug mode adjustments (TBD if we will keep all of this)

NoelStephensUnity committed:
Test

Updating tests for previous PR's changes.

NoelStephensUnity committed:
update and test

Slowly backing out some legacy changes and debug info from NetworkTransform
Updating some of the NetworkTransform tests to more recent changes.

NoelStephensUnity committed:
update

This applies delta position updates to HalfVector3 and includes adjustments to NetworkTransform to account for this.
This also increases half float quaternion precision.

NoelStephensUnity committed:
test manual

Adjustments for the HalfVector3 delta position update.

NoelStephensUnity committed:
test

Adjustments for HalfVector3 delta position.  We no longer need all of the extra garbage.

NoelStephensUnity committed:
update

Temporarily removing HalfVector4 until it uses a similar delta format as HalfVector3.

NoelStephensUnity committed:
update

This comes close to resolving issues with delta position for both server and owner authoritative modes.
It required some adjustments to the OnSynchronize in order to be able to accomplish this.
This version still has an outstanding issue with resetting interpolation to the current position.

NoelStephensUnity committed:
fix

This fixes the issue with interpolation not being set properly when using half float delta position.

NoelStephensUnity committed:
test manual

Updates to the manual nested network transform test using delta position with the HalfVector3 struct.

NoelStephensUnity committed:
test

Updates to the nested network transform test

NoelStephensUnity committed:
test

adjustments for the OnSynchronize update

NoelStephensUnity committed:
update

Working version of HalfVector3 that has a rolling delta and keeps precision without any adjustment up or down.
Almost working version of NetworkTransform that uses the delta as part of the synchronization process.
(still needs cleaning up)

NoelStephensUnity committed:
test manual

Updates to the delta position test.

NoelStephensUnity committed:
update

Cleaning up some of the code base, reducing the delta threshold to 16 (this might change, but seems pretty good).

NoelStephensUnity committed:
update

Finally figured out the issue with rotating in a circle and interpolation.
We were always using Lerp which would interpolate in a linear (i.e. straight) line from axis value A to axis value B.
This is a temporary adjustment that provides one with the ability to determine if they want to be lerping or slerping their interpolated position.
(Really, we need to provide an interface to users that will allow them to build their own interpolators.)

This also includes an update during synchronization to better determine if it needs to update for HalfVector3 (delta position) or not...still not tested with owner authoritative.

NoelStephensUnity committed:
test manual

Made some updates based on previous commits.

NoelStephensUnity committed:
update - wip

Server authority working with new timing update changes.

NoelStephensUnity committed:
update

This update has both owner and server authoritative modes working with HalfVector3 delta position updates.

NoelStephensUnity committed:
test manual

This update has come changes required by previous commit as well as some visual clean up and updates to the nested NetworkTransform manual test to be able to visually see that all automated player instances are in the same relative grid regions.

NoelStephensUnity committed:
test

Minor adjustments to this scene for visualization purposes.

NoelStephensUnity committed:
update

Migrating NetworkTick out of #if DEBUG_NETWORKTRANSFORM block as it is currently used when UseHalfFloatPrecision is enabled.

NoelStephensUnity committed:
style and minor update

Adding some more comments as well as minor tweak.

NoelStephensUnity committed:
update - precision adjustment

Adding the precision adjustment back in to avoid loss of precision over time.
Adding back the local previous position delta to avoid bouncing between imprecise values.

NoelStephensUnity committed:
test manual

Minor adjustments for testing purposes.

NoelStephensUnity committed:
update

removing as much debug related stuff as possible.
made some things accessible/public (TBD if we keep these changes).
adding the missed meta file for HalfVector4

NoelStephensUnity committed:
tests manual

Updating the manual tests used in the video as well as migrated most of the debug related code for NetworkTransform into IntegrationNetworkTransform

NoelStephensUnity committed:
test manual

missed the scenes in build update

NoelStephensUnity committed:
test

minor adjustments to integration tests due to previous commits.

NoelStephensUnity committed:
test

wip

NoelStephensUnity committed:
update - remove

Removing scripts no longer used.

NoelStephensUnity committed:
update - tick replacing sent time

We now only send the network tick as opposed to the double precision server time since authority updates state only on each new network tick.  While there might be a very small factional difference between the actual time and the non-authoritative calculated time from the tick, the over-all end result is that it appears to be just as smooth when interpolating (if not slightly smoother).
Also removing the compression folder left out in the previous commit.

NoelStephensUnity committed:
update

Compressing m_BitSet and m_NetworkTick.
Using references in a few more places.

NoelStephensUnity committed:
test manual

Updating manual tests for additive scene loading manual test with NetworkPrefabPool in order to be able to test turning on and off half float precision, quaternion synchronization, and quaternion compression.

NoelStephensUnity committed:
update - synchronize transform

This update allows users to determine if they want to synchronize the NetworkObject's associated transform when it is spawned.  This can help reduce the initial synchronization cost for things like in-scene placed NetworkObjects that have no visual components associated with them (i.e. managers etc) as well as "netcode aware" UI elements that have a pre-determined fixed position.

NoelStephensUnity committed:
update - quaternion compression

This just re-enables the quaternion compression option when serializing.

NoelStephensUnity committed:
test - manual Bandwidth Test

Added a simple manual test to determine NetworkTransform bandwidth costs for half float precision, quaternion synchronization and compression, and the like.

NoelStephensUnity committed:
update

updating editor build settings for the bandwidth test scene.

NoelStephensUnity committed:
update

Enabling serialization of half quaternion and quaternion compressed.
Added last sent size for stats purposes.

NoelStephensUnity committed:
update

Including scale when enabling half float precision.
A little bit of code clean up in NetworkTransformState.NetworkSerialize.

NoelStephensUnity committed:
update

WIP
Looking into owner-side initial rotation synchronization issue during synchronization.

NoelStephensUnity committed:
update

Fixes for issue with rotation update.  It was trying to convert from half float or decompress before reading the value during serialization.

NoelStephensUnity committed:
test

Updated to most recent NetworkTransform changes.

NoelStephensUnity committed:
update

re-enabling the compression of Bitset.

NoelStephensUnity committed:
update

renaming HalfVector3 to HalfVector3DeltaPosition since this version is really specific to handling delta position and not just a HalfVector3.

NoelStephensUnity committed:
update

This update includes:
NetworkTransform updates:
This updates some missed portions of the initial synchronization with scale and now, only if synchronizing will determine whether to synchronize the lossy or local scale.
Integrated the new HalfVector3 for scale synchronization.
*** This uses the new NetworkObject.WorldPositionStays method.***
Scale now uses BufferedLinearInterpolatorVector3 and BufferedLinearInterpolator<float> is no longer used.

Added HalfVector3:
Can be configured for axis specific synchronization,  decimal precision adjustment, and is now used to synchronize scale in NetworkTransform.

Added NetworkObject.WorldPositionStays accessor method for the cached WorldPositionStays value (needed by NetworkTransform).

NoelStephensUnity committed:
test

minor update of the player nested transform prefab

NoelStephensUnity committed:
update

Some fixes for integration tests.
InLocalSpace changes on authority side was not invoking a teleport.
Some non-authoritative side adjustments to prevent non-authority from moving synchronized axis.

NoelStephensUnity committed:
test

updates to some of the transform related integration tests.
includes a fix for TransformInterpolationTests failing on console.
minor clean up.

NoelStephensUnity committed:
update

Clean up and some minor fixes/improvements in various areas.
(i.e. only update interpolators that are being synchronized, improved non-authoritative enforcement of authoritative transform state, assure we are only checking portions of transform being synchronized on the authoritative side, improved error messages in places, etc.)

Updated XML API Documentation.

NoelStephensUnity committed:
test

Some updates to get all tests passing.

NoelStephensUnity committed:
test manual

Minor update to exclude toggles in additive scene loading and networkobject pool manual test.

NoelStephensUnity committed:
Merge branch 'develop' into refactor/quaternion-synchronization-network-transform

NoelStephensUnity committed:
style

fixing white spaces.

NoelStephensUnity committed:
style

removing Unity.Netcode using statement.

NoelStephensUnity committed:
test

fixing a mistake I made with checking who cannot commit to a transform

NoelStephensUnity committed:
style

formatting

NoelStephensUnity committed:
update

Setting some of the set accessors of NetworkTransformState properties to internal.
Adding additional public API xml documentation.

NoelStephensUnity committed:
style

Updating XML API documentation

NoelStephensUnity committed:
update

making HalfVector3DeltaPosition.CurrentBasePosition internal, added accessor method, adjusted/added some XML API documentation.

NoelStephensUnity committed:
fix

Fixing issue with HalfVector3 decimal precision calculation (wrong place to divide into 1).

NoelStephensUnity committed:
fix

Adding back the no longer used m_CachedIsServer and m_CachedNetworkManager properties since these were made protected originally.  :(

NoelStephensUnity committed:
style

updating comments as some of that would be confusing.

NoelStephensUnity committed:
update

removing bandwidth test from this PR and making it a separate PR.

NoelStephensUnity committed:
update

Testing using my original delta compression

NoelStephensUnity committed:
update

Minor adjustment to delta threshold for delta position before collapsing to base position.
Still WIP on delta compressor.

NoelStephensUnity committed:
Merge branch 'develop' into refactor/quaternion-synchronization-network-transform

NoelStephensUnity committed:
update

Merge modification.

NoelStephensUnity committed:
Merge branch 'develop' into feat/quaternion-synchronization-network-transform

NoelStephensUnity committed:
update

Minor update

NoelStephensUnity committed:
update

Removing some of the adjust up and down as it makes very little (if no under almost most scenarios) real difference.

NoelStephensUnity committed:
test manual

Some minor tweaks on value decimal place output for the delta position test.
Also as opposed to stopping immediately, it gradually stops.

NoelStephensUnity committed:
Merge branch 'develop' into feat/quaternion-synchronization-network-transform

NoelStephensUnity committed:
Merge branch 'develop' into feat/quaternion-synchronization-network-transform

NoelStephensUnity committed:
Merge branch 'develop' into feat/quaternion-synchronization-network-transform

NoelStephensUnity committed:
Merge branch 'develop' into feat/quaternion-synchronization-network-transform

NoelStephensUnity committed:
fix - merging parent dynamic under in scene placed

Just merging this fix ahead of time.

NoelStephensUnity committed:
update

Fixing merge issue with editor build settings.
Adding tool tips to additional properties of NetworkTransform.
Adding new properties to NetworkTransformEditor.

NoelStephensUnity committed:
Test Manual

Some fixes and updates for test project.

NoelStephensUnity committed:
Merge branch 'develop' into feat/quaternion-synchronization-network-transform

NoelStephensUnity committed:
update

Removing the delta position compressor as it is no longer used.

NoelStephensUnity committed:
update

Adding the Slerp property to the NetworkTransformEditor

NoelStephensUnity committed:
style

fixing white space issue.

samlucas-unity commented on com.unity.netcode.gameobjects/Components/HalfVector3.cs:
```diff
@@ -0,0 +1,141 @@
+        /// <param name="serializer"></param>
+        public void NetworkSerialize<T>(BufferSerializer<T> serializer) where T : IReaderWriter
+        {
+            if (m_HalfVector3AxisToSynchronize.X)
```
I'm guessing we are relying on both ends having same same m_HalfVector3AxisToSynchronize set and not explicitly synchronizing that state?

samlucas-unity commented on com.unity.netcode.gameobjects/Components/HalfVector3.cs:
```diff
@@ -0,0 +1,141 @@
+        /// <param name="serializer"></param>
+        public void NetworkSerialize<T>(BufferSerializer<T> serializer) where T : IReaderWriter
+        {
+            if (m_HalfVector3AxisToSynchronize.X)
```
I'm guessing we are relying on both ends having same same m_HalfVector3AxisToSynchronize set and not explicitly synchronizing that state?

NoelStephensUnity replied:
That is a good question.  Since the [axis synchronized are defined by NetworkTransform](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/0d9709acfa2cb8e3348d0b996b3c18adf7475daa/com.unity.netcode.gameobjects/Components/NetworkTransform.cs#L613) during serialization (and my mind was thinking about bandwidth optimizations at the time) it seemed like a waste of additional bandwidth to synchronize which axis were being synchronized each time it was serialized.  
Definitely open to discussing if we  want to somehow make it possible to synchronize this during serialization, but currently with half float precision we get roughly 25 bytes per full transform update (21 bytes with Quaternion compression) per instance.  For the most part those numbers are about as packed as they can get (give or take a couple of bits), adding any additional information (i.e. which axis to synchronize) would increase those numbers by 1 byte per instance.
However, since I made the struct public a user might want to use that for some other purpose than what it is being used for... so we could add an additional bool, via public get/set or constructor, that defaults to false and would serialize m_HalfVector3AxisToSynchronize first if set to true.  

Thoughts?

samlucas-unity replied:
Yeah I was thinking its just a visibility issue, its perfectly fine to assume they will be the same on both ends as long as we provide some method to detect it isn't, even if that's only in debug builds or with error checking enabled.

NoelStephensUnity replied:
Since authority dictates which axis are updated,  all to-be-synchronized axis are included. It only requires one of the to-be-synchronized axis to have a delta and then all to-be-synchronized axis are serialized and the corresponding NetworkTransfomState's HasPositionX, HasPositionY, HasPositionZ values are set respectively and serialized (write). 
 These values define which axis the non-authoritative side should expect.  So, the HalfVector3 or HalfVector3DeltaPosition initialize with that specific HalfVector3AxisToSynchronize setting prior to serialization (reading) on the non-authoritative side(s).

samlucas-unity commented on com.unity.netcode.gameobjects/Components/HalfVector3DeltaPosition.cs:
```diff
@@ -0,0 +1,204 @@
+    /// Half float precision <see cref="Vector3"/> that is used for delta position
+    /// synchronization.
+    /// </summary>
+    public struct HalfVector3DeltaPosition : INetworkSerializable
```
This might be more of an overall design thing but why doesn't the HalfVector3Delta offer the ability to only synchronize some axes like HalfVector3?

NoelStephensUnity replied:
You know, I started to think I had an answer to this but then realized you are making a good point.  Why not provide delta position for only a specific set of axis?  Let me mull over this today and if I don't find a good reason why (in the code) we wouldn't want to do this... then I think we should add this capability.  Good suggestion/catch! 👍

samlucas-unity commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -60,52 +64,81 @@ internal struct NetworkTransformState : INetworkSerializable
+            private const int k_PositionSlerp = 16;
 
-            private ushort m_Bitset;
+            private uint m_Bitset;
```
I like the BitGet/Set function change is there not a bitfield utility class somewhere which can do all of this?

NoelStephensUnity replied:
The most recent updates no longer require bit shifting.  👍

samlucas-unity commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -47,7 +48,10 @@ public class NetworkTransform : NetworkBehaviour
+        /// </summary>
+        public struct NetworkTransformState : INetworkSerializable
         {
             private const int k_InLocalSpaceBit = 0;
```
These don't seem to be used by index anywhere.  It might be better to just declare them as the shifted version upfront then you don't need to do that on every use.

NoelStephensUnity replied:
🐱 
You know... another really good point.  I think I fell into the trap of following the existing usage pattern...which it is nice to get a new set of eyes on this.  You are 100% correct and we could remove the BitGet and BitSet methods and just do the Boolean operations in the setters and getters... nice suggestion!  👍

samlucas-unity commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -114,167 +147,494 @@ internal bool HasPositionChange
+                if (HasScaleChange)
                 {
-                    serializer.SerializeValue(ref ScaleZ);
+                    if (UseHalfFloatPrecision)
```
It might be worth splitting out the serializer into a halfprecision version and a normal version and just select the version based on the flag since it applies as a whole.

NoelStephensUnity replied:
Yeah, I thought about this. If you review the NetworkTransform System Analysis document one more time you will see that when we implement the state processors this kind of change will be easier to implement.  Which that is another thing to contemplate... whether we want to use this PR as the base for that update and then just roll out all of the updates in one PR. On the other hand, this PR resolves several issues... so I have been on the fence about whether it would be better to have this as the "last update for NetworkTransform using this approach"... the biggest struggle I have been having with this PR is making the NetworkTransformState public...that makes it a long term commitment...
_(worth thinking about)_

NoelStephensUnity replied:
implemented.  👍

NoelStephensUnity committed:
test

Re-enabling half precision checks for NetworkTransformStateTests.

NoelStephensUnity committed:
update

Updating changelog
Minor adjustment to an unrequired virtual method.

NoelStephensUnity committed:
update and style

Updating comments.
Adding one missed entry to the changelog.

samlucas-unity commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -582,13 +1149,20 @@ internal NetworkTransformState ApplyLocalNetworkState(Transform transform)
         /// </summary>
-        private bool ApplyTransformToNetworkStateWithInfo(ref NetworkTransformState networkState, double dirtyTime, Transform transformToUse)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool ApplyTransformToNetworkStateWithInfo(ref NetworkTransformState networkState, ref Transform transformToUse, bool isSynchronization = false, ulong targetClientId = 0)
```
I can see why you wanted to split everything out in the new design. This component does so many things 😀

NoelStephensUnity replied:
Yep! 😸 
_(see previous comment)_

samlucas-unity commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -599,84 +1173,309 @@ private bool ApplyTransformToNetworkStateWithInfo(ref NetworkTransformState netw
+                }
+            }
+            else // If we are synchronizing then we need to determine which scale to use
+            if (isSynchronization && SynchronizeScale)
```
No need to check isSynchronization  here

NoelStephensUnity replied:
!doh!
I am going to blame that on a previous iteration where it did make sense... but now that you point that out and after reviewing... good catch!  👍

samlucas-unity commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -686,140 +1485,263 @@ private bool ApplyTransformToNetworkStateWithInfo(ref NetworkTransformState netw
-                if (SyncPositionY) { adjustedPosition.y = m_PositionYInterpolator.GetInterpolatedValue(); }
-                if (SyncPositionZ) { adjustedPosition.z = m_PositionZInterpolator.GetInterpolatedValue(); }
+                var interpolatedPosition = m_PositionInterpolator.GetInterpolatedValue();
+                if (SynchronizePosition)
```
Here we are checking the SynchronizePosition flag then selecting between half precision or not but in ApplyTransformToNetworkStateWithInfo the SynchronizePosition is checked after its checked not half precision.  If that correct logic?  I can't tell from just reading the diff if its correct or not so pointing it out incase it raises any flags.

NoelStephensUnity replied:
That should be nested...
```
                if (SynchronizePosition)
                {
                    var interpolatedPosition = m_PositionInterpolator.GetInterpolatedValue();
```
👍

samlucas-unity commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -1182,63 +2300,51 @@ private void SetStateServerRpc(Vector3 pos, Quaternion rot, Vector3 scale, bool
-            // If we are authority, update the authoritative state
-            if (CanCommitToTransform)
-            {
-                UpdateAuthoritativeState(transform);
```
I looks like the code for this function is now in OnUpdateAuthoritativeState, just checking its OK that this isn't called anymore.

NoelStephensUnity replied:
Yeah, this is intentional. 
_It used to be that authority would check every frame for deltas and then each new tick would send the final deltas from the previous tick. We would basically "collapse"/overwrite the deltas in-between ticks and doing all of that processing ended up being wasteful._  
Now, we just make authority subscribe to the `NetworkTickSystem.Tick` event and check for deltas once and if any are detected update the nonauthoritative instances with the values.

NoelStephensUnity committed:
update

HalfVector3DeltaPosition now configures itself based on the NetworkTransform's position axis selected for synchronization.
Includes some improvements in how axis are serialized.

NoelStephensUnity committed:
test

Includes several updates to NetworkTransformTests.
Includes updates and adjustments based on the previous update.

NoelStephensUnity committed:
update

Implementing PR suggestion with flags for NetworkTransformState

NoelStephensUnity committed:
test

Enabling verbose debug if NestedNetworkTransform fails.

NoelStephensUnity committed:
update

Adding additional check to make sure if SyncPosition axis value changes while using half precision (delta position) that if the delta on that axis exceeds the maximum delta that the NetworkObject is teleported on the tick this condition is detected.

NoelStephensUnity committed:
test manual

updates the delta position long duration test for precision drift so that it randomly selects x,y,z axis for direction and synchronization in order to determine if enabling and disabling which axis should be synchronized during runtime can cause issues.
(still need to add an integration test for this too)

NoelStephensUnity committed:
update and style

Added accessor methods for NetworkTransformState to get position and scale.
Updated the rotation accessor method in NetworkTransformState.
Minor update to NetworkTransformState.NetworkSerialize to keep the isWriting usage consistent.
Updated a bunch of XML API entries.

NoelStephensUnity committed:
Merge branch 'develop' into feat/quaternion-synchronization-network-transform

NoelStephensUnity committed:
changelog fix

migrating last PR's changelog entry into  the Unreleased section

ShadauxCat commented on com.unity.netcode.gameobjects/Components/HalfVector3.cs:
```diff
@@ -0,0 +1,114 @@
+                if (m_HalfVector3AxisToSynchronize.SyncAxis[i])
+                {
+                    var axisValue = Axis[i];
+                    serializer.SerializeValue(ref axisValue);
```
You could get even more optimization out of this by using `BytePacker.WriteValuePacked(serializer.GetWriter(), Axis[i])` and `ByteUnpacker.ReadValuePacked(serializer.GetReader(), out Axis[i])`

NoelStephensUnity replied:
Would like to but the ushort values are half floats which are already as packed as they can get.

ShadauxCat commented on com.unity.netcode.gameobjects/Components/HalfVector4.cs:
```diff
@@ -0,0 +1,68 @@
+
+        public void NetworkSerialize<T>(BufferSerializer<T> serializer) where T : IReaderWriter
+        {
+            serializer.SerializeValue(ref X);
```
Same comment as HalfVector3 - pack/unpack

NoelStephensUnity replied:
Same as previous response about the ushort values being half floats

NoelStephensUnity committed:
Merge branch 'develop' into feat/quaternion-synchronization-network-transform

NoelStephensUnity committed:
update

Adding controls to randomly switch axis and reset them all to true when stopping.

NoelStephensUnity committed:
update

Euler rotation can be synched per axis when half precision is enabled but not when quaternion synchronization is enabled.
Quaternion lerps when half precision is enabled and slerps when full precision is enabled.

NoelStephensUnity committed:
test

Updating tests to take into account per axis Euler rotation when half precision is enabled.
Improved and fixed the Euler approximation integration test methods.

NoelStephensUnity committed:
Merge branch 'develop' into feat/quaternion-synchronization-network-transform

NoelStephensUnity committed:
fix

Wrapping the transport shutdown process with some try catch statements to avoid an infinite loop shutdown in the event there is an exception thrown when trying to shutdown transport.

NoelStephensUnity committed:
Merge branch 'develop' into feat/quaternion-synchronization-network-transform

NoelStephensUnity committed:
Merge branch 'develop' into feat/quaternion-synchronization-network-transform

NoelStephensUnity committed:
update

correcting changelog merge issues.

NoelStephensUnity committed:
Merge branch 'develop' into feat/quaternion-synchronization-network-transform

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -1085,13 +1085,22 @@ private void ClearSendQueuesForClientId(ulong clientId)
         private void FlushSendQueuesForClientId(ulong clientId)
         {
-            foreach (var kvp in m_SendQueue)
+            // NSS: In the off chance this fails, log an exception
```
Will be removed when PR-2433 is merged.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -1200,7 +1200,19 @@ internal void ShutdownInternal()
             {
                 // Client only, send disconnect to server
-                NetworkConfig.NetworkTransport.DisconnectLocalClient();
+                // NSS: In the event this fails or throws and exception
```
Will be removed when PR-2433 is merged.

NoelStephensUnity committed:
Merge branch 'develop' into feat/quaternion-synchronization-network-transform

NoelStephensUnity committed:
revert

Reverting out the changes to UnityTransport as this was fixed in PR-2433.

ShadauxCat commented on com.unity.netcode.gameobjects/Components/HalfVector3.cs:
```diff
@@ -0,0 +1,114 @@
+namespace Unity.Netcode.Components
+{
+    /// <summary>
+    /// Half float precision <see cref="Vector3"/>
```
Based on our previous conversation, could you make a comment here about how the ushorts in this class are binary reinterpretations of half-floats because C# doesn't support them at the language level? Also in the other cases where you use this technique. Would be good to know for anyone who comes and looks at it later.

ShadauxCat commented on com.unity.netcode.gameobjects/Components/HalfVector3.cs:
```diff
@@ -0,0 +1,114 @@
+        /// </summary>
+        /// <param name="vector3">the <see cref="Vector3"/> to convert to a half precision <see cref="Vector3"/></param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void FromVector3(ref Vector3 vector3)
```
Nitpicky, but this method name feels misleading to me. I generally expect `X.FromY()` to be a static method that returns a new `X`. Having it be an instance method that changes the instance and returns `void` surprised me.

Could I suggest `AssignFromVector3()`? Or maybe even just `Assign()`? `halfFloat.Assign(myVector3)` seems intuitive to me.

NoelStephensUnity replied:
I think I like your `halfFloat.Assign(myVector3)` better too!

ShadauxCat commented on com.unity.netcode.gameobjects/Components/HalfVector3AxisToSynchronize.cs:
```diff
@@ -0,0 +1,41 @@
+        /// <param name="z">when <see cref="true"/> the z-axis will be serialized</param>
+        public HalfVector3AxisToSynchronize(bool x = true, bool y = true, bool z = true)
+        {
+            SyncAxis = new bool[3];
```
This will allocate managed memory. Do we care about that? Could we make this a NativeArray instead, or just three bool values?

ShadauxCat replied:
Actually even better would be `FixedList32Bytes`, which gives the list interface in a fixed memory so it doesn't require any allocators or adding IDisposable. Slightly too large for three bools so it wouldn't have a length of 3 and you'd have to account for that... having just bool x, bool y, bool z would be the best option in terms of memory usage and allocation behavior.

ShadauxCat replied:
Another option would be a simple class like this:

```csharp
public struct Vector3T<T> where T: unmanaged
{
    // No other members, ensures array-style access works
    public T X;
    public T Y;
    public T Z;
    
    // Property, not value
    public int Length => 3;
    
    public unsafe T this[int index]
    {
        get 
        {
            fixed (Vector3T<T>* ptr = &this)
            {
                return ((T*)ptr)[index];
            }
        }
        set 
        {
            fixed (Vector3T<T>* ptr = &this)
            {
                ((T*)ptr)[index] = value;
            }
        }
    }
}
```

If you wanted, you could even give that an interface:

```csharp
public interface VectorNT<T>
{
    int Length { get; };
    T this[int index] { get; set };
}
```

Then you could make Vector2T, Vector3T, Vector4T, and pass them around using Length and array-access.

That'd let you keep using the programming style you're using, while avoiding both managed and unmanaged memory allocation, keeping everything in a fixed-size struct.

NoelStephensUnity replied:
Nice! I will look at implementing this.  I could even use this for the bools issue as well.  👍

NoelStephensUnity replied:
of course... I will replace:
`FastBufferWriter* ptr` with `Vector3T<T>*`

But I get the general direction you were heading with that.

ShadauxCat replied:
Oh... yeah... oops. I wanted to test my syntax was correct so I just wrote some test code inside the class I happened to have open and forgot to adjust it.

ShadauxCat commented on com.unity.netcode.gameobjects/Components/HalfVector3.cs:
```diff
@@ -0,0 +1,114 @@
+        /// valid values range from 0 to 4 decimal places to adjust up and back down (1000 to 1 : 1 to 0.001).
+        /// </remarks>
+        /// <param name="vector3">the vector3 to initialize the HalfVector3 with</param>
+        public HalfVector3(Vector3 vector3, HalfVector3AxisToSynchronize halfVector3AxisToSynchronize)
```
Could we have another pair of constructors with the second parameter defaulting to (true, true, true)?

```csharp
public HalfVector3(Vector3 vector3) :
    this(vector3, new HalfVector3AxisToSynchronize())
{
}
```

`HalfVector3AxisToSynchronize` is very descriptive but also a lot to type if users are using the default values anyway.

NoelStephensUnity replied:
I think if I were to rename that AxisToSynchronize it would help with the name, but it would be easy to have a default of all axis too.  👍

ShadauxCat commented on com.unity.netcode.gameobjects/Components/HalfVector3DeltaPosition.cs:
```diff
@@ -0,0 +1,214 @@
+        /// </summary>
+        public ushort Z => Axis[2];
+
+        public ushort[] Axis;
```
Same comment re: this being `NativeArray<ushort>`, `FixedList32Bytes<ushort>`, or just three ushort values, instead of managed memory.

(Edit: Or `Vector3T<ushort>` - see my comment above.)

NoelStephensUnity replied:
👍

ShadauxCat commented on com.unity.netcode.gameobjects/Components/HalfVector3.cs:
```diff
@@ -0,0 +1,114 @@
+        /// </summary>
+        public ushort Z => Axis[2];
+
+        public ushort[] Axis;
```
Actually, also same comment here as I made below... `NativeArray<ushort>`, `FixedList32Bytes<ushort>`, or just three ushort values, instead of managed memory.

(Edit: Or `Vector3T<ushort>` - see my comment below.)

ShadauxCat commented on com.unity.netcode.gameobjects/Components/HalfVector4.cs:
```diff
@@ -0,0 +1,68 @@
+{
+    internal struct HalfVector4 : INetworkSerializable
+    {
+        public ushort X;
```
I'm curious now if there's a reason you used ushort[] for HalfVector3, but just used four values for HalfVector4.

NoelStephensUnity replied:
I think all of these need a face lift actually after reading some of your comments. 😸

NoelStephensUnity committed:
Merge branch 'develop' into feat/quaternion-synchronization-network-transform

NoelStephensUnity committed:
update

Updating the vector implementations to reduce memory allocation.
Renaming some structs and methods for better clarity.
Updated XML API for clarity and formatting.

NoelStephensUnity committed:
Merge branch 'develop' into feat/quaternion-synchronization-network-transform

NoelStephensUnity committed:
style

Updating some wording in the XML API.

NoelStephensUnity committed:
style

adding suggested comments about leaving remarks regarding the contents of the ushort values are used to store the 16bits of the half float values.

wackoisgod commented:
I am curious why these new types were added when Unity already has these types in the Unity Mathematics package? Like many things within NGO it seems that you just end up creating your own version of a thing? Why not just make the types in a package that is used all over the place support serialization? and furthermore are compatible with Engine Types when required?

NoelStephensUnity commented:
@wackoisgod 
Which types are you referring to?

wackoisgod commented:
> 

That would be correct; I just think introducing another new type seems kind of wasteful and confusing to people because now they can use the Unity.Mathmatics package which converts to Unity Engine types where required but then now they may have to convert to these NGO types when they want to use them on the network. I just don't see a lot of value in creating another type.


I think VectorT is not a great type either because while yes it can be limited to an unmanaged type its really only value is primitive  number types but nothing stops someone from putting any sort of struct in the type.

NoelStephensUnity commented:
@wackoisgod 
I see your point. We still have the [release/1.3.0](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/tree/release/1.3.0) update to push out before any of the changes that occurred after the v1.3.0 branch...so there is plenty of time to look into any fine tuning/re-factoring.  All input is 100% welcome (_especially yours_).
The Unity.Mathmatics package would need to be added to the assemblies referenced in all NGO asmdef files (currently it is not referenced) which would make NGO dependent upon that assembly.  I will most definitely pass it by @ThusSpokeNomad to get his take on adding that as a dependency.

Please always feel free to bring up anything you see that needs a second pass "re-think" as I highly value your insights into things of this very nature and the more eyes on things the better (imo).  👍

ThusSpokeNomad commented:
We already indirectly depend on `Unity.Mathematics` package, so it'd be just fine :)
<img width="1635" alt="image" src="https://user-images.githubusercontent.com/12574651/225283790-8358e4bf-923e-45c5-bb7e-58fc8ab8b473.png">
Would this depgraph be useful? @NoelStephensUnity

NoelStephensUnity commented:
@ThusSpokeNomad Very cool graph...
@wackoisgod 
#2456 implements the refactoring with Unity.Mathematics types.
Most likely I will need to update the FastBufferReader/Write with all types included in Unity.Mathematics, so the PR is still in draft form (_just doing a pass on desktop platforms to make sure nothing else is impacted_).
